### PR TITLE
[SPARK-42521][SQL] Add NULLs for INSERTs with user-specified lists of fewer columns than the target table

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

Add NULLs for INSERTs with user-specified lists of fewer columns than the target table.

This is done by updating the semantics of the `USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES` SQLConf to only apply for INSERTs with explicit user-specific column lists, and changing it to true by default.

### Why are the changes needed?

This behavior is consistent with other query engines.

### Does this PR introduce _any_ user-facing change?

Yes, per above.

### How was this patch tested?

Unit test coverage in `InsertSuite`.

### CODE REVIEW

**Comment by dtenedor** [comment]
Hi @gengliangwang 👋

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -3126,13 +3126,12 @@ object SQLConf {
...
-        "explicit default values to behave as if they had specified DEFAULT NULL instead. " +
-        "For example, this allows most INSERT INTO statements to specify only a prefix of the " +
-        "columns in the target table, and the remaining columns will receive NULL values.")
+      .doc("When true, and DEFAULT columns are enabled, allow INSERT INTO commands with user-" +
+        "specified lists of fewer columns than the target table to behave as if they had " +
+        "specified DEFAULT for all remaining columns instead, in order.")
       .version("3.4.0")
       .booleanConf
-      .createWithDefault(false)
+      .createWithDefault(true)
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3134]: Let's update `sql-migration-guide` as well

**Comment by gengliangwang** [comment]
LGTM except one comment and the pending test failures.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3134]: Sure, this is done!

File: docs/sql-migration-guide.md [hunk 2]
```diff
@@ -24,6 +24,7 @@ license: |
 
 ## Upgrading from Spark SQL 3.3 to 3.4
   
+  - Since Spark 3.4, INSERT INTO commands with explicit column lists comprising fewer columns than the target table will automatically add the corresponding default values for the remaining columns (or NULL for any column lacking an explicitly-assigned default value). In Spark 3.3 or earlier, these commands would have failed returning errors reporting that the number of provided columns does not match the number of columns in the target table.
```
Comment (gengliangwang) [inline docs/sql-migration-guide.md:27]: Let's mention users can disable `spark.sql.defaultColumn.useNullsForMissingDefaultValues` to restore the previous behavior?

Comment (dtenedor) [reply docs/sql-migration-guide.md:27]: Good idea, done!

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master/3.4
This makes the column default feature simpler and more reasonable.
cc @xinrong-meng

**Pull request closed without merging**